### PR TITLE
Age 4 configuration

### PR DESCRIPTION
**Pull request by julien-devatom**
### Description of change
- first step of the #87 
- ext steps needs a snapshot block, one hour before the beginning of the epoch

### Pull-Request Checklist

- [x] Code is up-to-date with the `main` branch
- [x] `yarn lint` passes with this change
- [x] `yarn test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [x] The new commits follow conventions outlined in the [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

File: src/ages/distributions/ageFourDistribution/ageFourDistribution.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import marketsRepartition from "./marketsRepartition";
+import { weightedDistribution } from "../weightedDistribution";
+import { PercentMath } from "@morpho-labs/ethers-utils/lib/maths";
+
+export const ageFourDistribution = async (
+  ageConfig: AgeDistribution,
+  { finalTimestamp, initialTimestamp, number, snapshotBlock, totalEmission }: EpochConfig,
+  provider?: providers.Provider
+) => {
+  if (!snapshotBlock) throw Error(`Cannot distribute tokens for epoch ${number}: no snapshotBlock`);
```
Comment (tomrpl) [inline src/ages/distributions/ageFourDistribution/ageFourDistribution.ts:14]: Do we name age4 and epoch nb 10 the same thing?

**Review by tomrpl — APPROVED** [review]

File: src/ages/distributions/ageFourDistribution/ageFourDistribution.ts (hunk 1 above)
Comment (julien-devatom) [reply src/ages/distributions/ageFourDistribution/ageFourDistribution.ts:14]: Yes, the refactor to remove the idea of epochs includes a lot of changes. So the repo is going to keep 1 epoch per age, named `ageX`

**Review by mlgarchery — APPROVED** [review]

File: src/ages/distributions/weightedDistribution.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      const address = MARKETS[symbol as keyof typeof MARKETS];
+
+      if (!address) throw Error(`Cannot distribute MORPHO: unknown market ${symbol}`);
+      const marketData = markets.find((market) => market.address.toLowerCase() === address.toLowerCase());
+      if (!marketData) throw Error(`Cannot distribute MORPHO: no market data for ${symbol}`);
+
+      const { morphoSupplyMarketSize, morphoBorrowMarketSize, p2pIndexCursor } = marketData;
+
+      const distribution = getWeight(symbol);
+      const total = morphoSupplyMarketSize.add(marketData.morphoBorrowMarketSize);
```
Comment (oumar-fall) [inline src/ages/distributions/weightedDistribution.ts:26]:
```suggestion
      const total = morphoSupplyMarketSize.add(morphoBorrowMarketSize);
```

File: src/ages/distributions/weightedDistribution.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      const supply = morphoSupplyMarketSize.mul(distribution).div(total);
+      const supplyRate = supply.div(duration);
+      const borrow = morphoBorrowMarketSize.mul(distribution).div(total);
+      const borrowRate = borrow.div(duration);
+      const marketEmission = supply.add(borrow);
```
Comment (oumar-fall) [inline src/ages/distributions/weightedDistribution.ts:31]: this is equals to `distribution` i guess

File: src/ages/distributions/ageFourDistribution/ageFourDistribution.ts (hunk 1 above)
Comment (ghivert) [inline src/ages/distributions/ageFourDistribution/ageFourDistribution.ts:14]: Add `fallbackSnapshotBlock`?

File: src/ages/distributions/ageThreeDistribution.ts [hunk 4]
```diff
@@ -37,43 +18,22 @@ export const ageThreeDistribution = async (
...
-          borrow,
-          borrowRate,
-          marketEmission,
-          morphoBorrow: morphoBorrowMarketSize,
-          morphoSupply: morphoSupplyMarketSize,
-          p2pIndexCursor,
-        } as MarketEmission,
-      ];
-    })
+  const getWeight = (symbol: string) => {
```
[the remark this answers is not in the document]
Comment (oumar-fall) [reply src/ages/distributions/ageThreeDistribution.ts:23]: or `getMarketEmission`

**Review by ghivert — APPROVED** [review]

File: src/ages/distributions/ageFourDistribution/ageFourDistribution.ts (hunk 1 above)
Comment (julien-devatom) [reply src/ages/distributions/ageFourDistribution/ageFourDistribution.ts:14]: I'll do in another PR

File: src/ages/distributions/ageThreeDistribution.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (julien-devatom) [reply src/ages/distributions/ageThreeDistribution.ts:23]: Youre right, that was the first implementation (returining the weight between 0 and 1), but I missed to rename it 👍🏼

**Pull request merged**